tools/ctl/ide/commands/commands.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2025 Google LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package commands
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/google/oss-rebuild/pkg/rebuild/rebuild"
+	"github.com/google/oss-rebuild/pkg/rebuild/schema"
+	"github.com/google/oss-rebuild/tools/benchmark"
+	"github.com/google/oss-rebuild/tools/ctl/diffoscope"
+	"github.com/google/oss-rebuild/tools/ctl/ide/assistant"
+	"github.com/google/oss-rebuild/tools/ctl/ide/chatbox"
+	"github.com/google/oss-rebuild/tools/ctl/ide/choice"
+	"github.com/google/oss-rebuild/tools/ctl/ide/details"
+	"github.com/google/oss-rebuild/tools/ctl/ide/modal"
+	"github.com/google/oss-rebuild/tools/ctl/ide/rebuilder"
+	"github.com/google/oss-rebuild/tools/ctl/ide/tmux"
+	"github.com/google/oss-rebuild/tools/ctl/localfiles"
+	"github.com/google/oss-rebuild/tools/ctl/rundex"
+	"github.com/pkg/errors"
+	"github.com/rivo/tview"
+	"gopkg.in/yaml.v3"
+)
+
+type RebuildCmd struct {
+	Short  string
+	Hotkey rune
+	Func   func(context.Context, rundex.Rebuild)
+}
+
+type GlobalCmd struct {
+	Short  string
+	Hotkey rune
+	Func   func(context.Context)
+}
+
+// A modalFnType can be used to show an InputCaptureable. It returns an exit function that can be used to close the modal.
+type modalFnType func(modal.InputCaptureable, modal.ModalOpts) func()
+
+type Commander struct {
+	app         *tview.Application
+	rb          *rebuilder.Rebuilder
+	modalFn     modalFnType
+	butler      localfiles.Butler
+	asst        assistant.Assistant
+	buildDefs   rebuild.LocatableAssetStore
+	dex         rundex.Reader
+	benches     benchmark.Repository
+	rebuildCmds []RebuildCmd
+	globalCmds  []GlobalCmd
+}
+
+func NewCommander(app *tview.Application, rb *rebuilder.Rebuilder, modalFn modalFnType, butler localfiles.Butler, asst assistant.Assistant, buildDefs rebuild.LocatableAssetStore, dex rundex.Reader, benches benchmark.Repository) (*Commander, error) {
```

Comment:
okay so overall, I think this is the right direction but we should split out the commands themselves into a separate constructor so the Commander's state is only the commands themselves. it then acts as a "registry" that can be initialized from ui or explorer while the commands can be more closely associated with the "component" that defines them. (the fudge here is rebuilds which are numerous enough that using real instantiated components for them is performance-prohibitive so special-casing them here is preferable.)